implement locate and updatedb

Qelxiros opened the pull request:
closes #60

Given that locate has largely been replaced by (m|p)locate, I was unable to find a version of locate packaged for my system (latest Arch Linux), and the packaged version of updatedb does not use the LOCATE02 db format. When running updatedb from source, I encountered errors that were not immediately fixable. As such, this implementation is largely untested, but it's better to have something here to be improved upon.

Qelxiros committed:
implement locate

sylvestre commented on src/locate/main.rs:
```diff
@@ -0,0 +1,5 @@
+fn main() {
```
please add the license header on all files :)

sylvestre commented on src/locate/mod.rs:
```diff
@@ -0,0 +1,433 @@
+use std::{
```
don't hesitate to add some comments :)

sylvestre commented:
cool, thanks :)
a bunch of tasks are failing

sylvestre commented:
add please add tests :)

Qelxiros committed:
locate fixes, implement updatedb, add tests

Qelxiros commented:
A couple notes:
- User flags are ignored for updatedb. GNU/BSD/Apple implementations all use a shell script, so they have direct access to su. I'm not sure if there's a better solution from rust, but I left it alone for now. 
- Performance is pretty significantly worse than GNU for updatedb. It might be a problem with find, since updatedb delegates most of the work to it. Locate is about on par with GNU for the one test that I did. I'd love another set of eyes to look for improvements.
- I'll fix coverage issues tomorrow.

sylvestre commented on src/db_tests.rs:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2017 Google Inc.
```
```suggestion
```

sylvestre commented on src/db_tests.rs:
```diff
@@ -0,0 +1,79 @@
+// Copyright 2017 Google Inc.
+//
```
and tests should go in tests/ :)

Qelxiros committed:
move test file, unix-gate locate

Qelxiros committed:
improve coverage

Qelxiros committed:
handle windows better

Qelxiros committed:
rename db

Qelxiros committed:
fix tests

Qelxiros committed:
improve coverage

Qelxiros committed:
remove files that broke windows ci, updatedb: fix large negative deltas

Qelxiros committed:
improve coverage, handle invalid databases

Qelxiros committed:
add tests

Qelxiros committed:
add test

Qelxiros committed:
fix test for coverage

sylvestre commented on Cargo.toml:
```diff
@@ -11,14 +11,18 @@ authors = ["uutils developers"]
 nix = { version = "0.30", features = ["fs", "user"] }
 argmax = "0.3.1"
+itertools = "0.14.0"
+quick-error = "2.0.1"
```
please switch to thiserror
quick-error hasn't been maintained for a while:
https://crates.io/crates/quick-error

Qelxiros replied:
Will do. Worth noting that `uucore` explicitly recommends `quick-error` (https://github.com/uutils/coreutils/blob/main/src/uucore/src/lib/mods/error.rs#L156-L157) and several coreutils use it over `thiserror`. I'd be happy to open an issue/PR over there if you'd like to migrate to `thiserror` there as well.

sylvestre replied:
Yes please :)
I was going to open good first bugs for this!

sylvestre commented on src/locate/main.rs:
```diff
@@ -0,0 +1,17 @@
+// Copyright 2017 Google Inc.
```
please remove this line

sylvestre commented on src/locate/main.rs:
```diff
@@ -0,0 +1,17 @@
+
+#[cfg(windows)]
+fn main() {
+    println!("locate is unsupported on Windows");
```
maybe add why ? :)

sylvestre commented on Cargo.toml:
```diff
@@ -11,14 +11,18 @@ authors = ["uutils developers"]
 argmax = "0.3.1"
+itertools = "0.14.0"
+quick-error = "2.0.1"
+uutests = "0.0.30"
```
do you use any of the uutests functions?

sylvestre commented on tests/db_tests.rs:
```diff
@@ -0,0 +1,198 @@
+// Copyright 2017 Google Inc.
```
it would be nice to check the output of the commands and use uutests functions for that

Qelxiros replied:
the uutests functions really want a multicall binary like coreutils, which isn't what we have here. I'd be interested in doing a larger refactor to bring findutils more in line with coreutils, but that probably shouldn't be this pr. should I just remove uutests as a dependency (to address your previous comment) and add uutests everywhere once this pr lands?

sylvestre replied:
Yes please :)

Qelxiros committed:
requested changes

Qelxiros committed:
fix --non-existing

sylvestre commented on src/updatedb/mod.rs:
```diff
@@ -0,0 +1,335 @@
+// Copyright 2017 Google Inc.
```
Same, please remove that line and the others

Qelxiros replied:
You asked me in https://github.com/uutils/findutils/pull/536#discussion_r2041546326 to add the license header. Should I just be removing the Google line? If so, the most recent commit should be good. If I'm misunderstanding, let me know.

Qelxiros committed:
remove license lines

Code7R commented:
Is there a reason for the kill switch which blocks operation on Windows?

Proper locate tool may be really interesting in the Windows world, where one would not have to install mingw or cygwin environments.

Qelxiros commented:
@Code7R The current implementation of `locate` relies on the UNIX-only `OsStr::from_bytes`. It might be possible to work around this on Windows with `OsString::from_wide`, but that at least requires a new allocation, and I'm not well-versed enough with Windows encodings to confidently identify how to properly convert from the `&[u8]` (read in from the db) to `&[u16]` (required by `from_wide`). As for `updatedb`, There's nothing explicitly preventing it from running on Windows, but I can't say with certainty that it doesn't make any UNIX-only assumptions.

Qelxiros commented:
@sylvestre any updates here?

sylvestre commented:
sorry, let me have a look
i just rebased your pr

sylvestre committed:
address review: move DB fixtures, drop tree-dirtying test fixture, style cleanups

- move invalid_db/old_db/test_data_db from repo root to test_data/db/
- remove add_special_files fixture that wrote into the tracked tree
- replace clap-default unwrap()s and the match_entry CString unwrap
- justify unreachable! branches; drop writeln!(...).unwrap()
- remove #[allow(dead_code)] and unused localuser/netuser Config fields

sylvestre committed:
fix locate existence/symlink logic, regex errors, updatedb expression building

- locate -e/-E: check the full path with correct follow semantics (a broken
  symlink is non-existent under -L/--follow, existent under -P/--nofollow);
  previously the follow flag was paired with the no-follow probe and basename
  mode checked the wrong path
- locate --regex: report an uncompilable pattern as an error instead of
  silently dropping it (From -> TryFrom, new Error::InvalidRegex)
- updatedb: build the find argv as discrete tokens rather than a formatted
  string that gets re-split on whitespace; an empty prunefs/prunepaths can no
  longer produce an invalid expression (empty '( )' group)
- add unit tests for path_exists and integration tests for the above
